infrastructure/comms/src/connection/peer_connection/worker.rs

```diff
@@ -0,0 +1,320 @@
+            Ok(_) => {
+                // The loop exited cleanly.
+                match *lock {
+                    // The connection is still in a connected state, transition to Shutdown
+                    PeerConnectionState::Connected(_) | PeerConnectionState::Connecting(_) => {
+                        *lock = PeerConnectionState::Shutdown;
+                    },
+                    // Connection is in some other state, the loop exited without error
+                    // so we won't change the state to preserve failed or disconnected states.
+                    _ => {},
+                }
+            },
+            Err(err) => {
+                // Loop failed, update the connection state to reflect that
+                *lock = match err {
+                    ConnectionError::PeerError(err) => PeerConnectionState::Failed(err),
+                    _ => PeerConnectionState::Failed(PeerConnectionError::UnexpectedConnectionError),
+                };
+            },
+        }
+
+        Ok(())
+    }
+
+    /// The main loop for the worker. This is where the work is done.
+    /// The required connections are set up and messages processed.
+    fn main_loop(&mut self) -> Result<()> {
+        let monitor = self.connect_monitor()?;
+        let peer_conn = self.establish_peer_connection()?;
+        let consumer = self.establish_consumer_connection()?;
+
+        loop {
+            if let Some(msg) = self.receive_control_msg()? {
+                match msg {
+                    ControlMessage::Shutdown => break Ok(()),
+                    ControlMessage::SendMsg(frames) => {
+                        let payload = self.create_payload(frames)?;
+                        peer_conn.send(payload)?;
+                    },
+                    ControlMessage::Pause => {
+                        self.paused = true;
+                    },
+                    ControlMessage::Resume => {
+                        self.paused = false;
+                    },
+                }
+            }
+
+            if !self.paused {
+                self.forward_frames(&peer_conn, &consumer)?;
+            }
+
+            if let Ok(event) = monitor.read(1) {
+                self.handle_socket_event(event)?;
+            }
+        }
+    }
+
+    /// Handles socket events from the ConnectionMonitor. Updating connection
+    /// state as necessary.
+    fn handle_socket_event(&mut self, event: SocketEvent) -> Result<()> {
+        use SocketEventType::*;
+
+        match event.event_type {
+            Disconnected => {
+                let mut lock = acquire_write_lock!(self.connection_state)?;
+                *lock = PeerConnectionState::Disconnected;
+            },
+            Listening | Accepted | Connected => {
+                self.retry_count = 0;
+                let mut lock = acquire_write_lock!(self.connection_state)?;
+                match *lock {
+                    PeerConnectionState::Connecting(ref thread_ctl) => {
```

Comment:
Can we guarantee that the `event_type` is "Connected"? 
Is it definitely impossible for `event_type` to be `Accepted` when PeerConnection state is currently `Connected`? No race conditions, messages getting delayed on the wire etc that could result in this?

Or is it by design that a `Listening` or `Accepted` event stick us in the `Connected` state?

I think a diagram of the FSM would be helpful here

Reply:
This code has three branches, if currently `Connecting` transition to `Connected`; if currently `Connected` do nothing (idempotent, which I believe handles the case you mentioned) or otherwise, it's an error and the connection will transition to a `Failed` state. If it is possible, I don't think it would be a problem (?)

I'll removed `Listening` and update tests from here though, as that is `Connecting` more than `Connected`

Reply:
On second thoughts, the reason `Listening` was in there was because for `Inbound` and `Outbound` connections you can use `wait_connected_or_failure` after starting the PeerConnection: 

```rust
let conn = PeerConnection::new()...;
conn.start();
conn.wait_connected_or_failure()
```
If we don't update the state for `Listening`, `wait_connected_or_failure` will timeout for `Inbound` connections. This is because you will only get an `Accepted` event once the remote peer connects - so I'm not sure what's best here as the semantics of `Inbound` and `Outbound` are slightly different. `Connected` could mean that the connection is connected to a peer (`Outbound`) or ready to be connected to (`Inbound`) - or, for `Inbound` you should use `wait_accepted_or_failure` and we add another "active" connection state `Accepted(thread_ctl)` - for the purposes of the `ConnectionManager` I think the former will work and make working with PeerConnections the same for `Inbound` and `Outbound` - tricky!

Reply:
Hence my question -- I wanted to make sure you've thought about all the edge cases, because the code seemed to be making some assumptions that I wanted clarification on. Intuitively, I only expect to reach the "Connected" state on a "Connected" event, unless there's some doc / explanation describing otherwise.